### PR TITLE
Show impl<T> and explain it for trait bounds and operators

### LINKED ISSUE #26991: impl<T> is not introduced in the book
There's only a few references to `impl<` in TRPL:

1. src/doc/trpl/deref-coercions.md
2. src/doc/trpl/lifetimes.md
3. src/doc/trpl/unsized-types.md

None of these really mention that you use `impl<T>` to declare a typename that will participate in either the trait or item involved in the implementation.

### PR DESCRIPTION
**Pull request by lastorset**
I also included some smaller trait-related changes.

Fixes #26991.

r? @shepmaster 
r? @steveklabnik

### CODE REVIEW

File: src/doc/trpl/generics.md [hunk 1]
```diff
@@ -57,7 +57,7 @@ enum Result<T, E> {
 ```
 
 This type is generic over _two_ types: `T` and `E`. By the way, the capital letters
-can be any letter you’d like. We could define `Result<T, E>` as:
+can be any letter or word you’d like. We could define `Result<T, E>` as:
```
Comment (steveklabnik) [inline src/doc/trpl/generics.md]: While this is true, type parameters are virtually always a single letter.

File: src/doc/trpl/generics.md [hunk 2]
```diff
@@ -101,11 +101,6 @@ fn takes_two_things<T, U>(x: T, y: U) {
 }
 ```
 
-Generic functions are most useful with ‘trait bounds’, which we’ll cover in the
-[section on traits][traits].
-
-[traits]: traits.html
-
```
Comment (steveklabnik) [inline src/doc/trpl/generics.md]: why is this being removed?

File: src/doc/trpl/generics.md [hunk 3]
```diff
@@ -122,3 +117,28 @@ let float_origin = Point { x: 0.0, y: 0.0 };
...
+
+When you want to add an implementation for the generic struct, you just
+declare the type parameter after the `impl`:
+
+```rust
+# struct Point<T> {
+#     x: T,
+#     y: T,
+# }
+
```
Comment (steveklabnik) [inline src/doc/trpl/generics.md]: this newline needs to go away or we'll get a blank line in the output :)

File: src/doc/trpl/generics.md (hunk 2 above)
Comment (steveklabnik) [inline src/doc/trpl/generics.md]: oh i see, you moved it below

File: src/doc/trpl/operators-and-overloading.md [hunk 4]
```diff
@@ -81,3 +81,55 @@ will let you do this:
...
+    };
+
+    println!("Area of s: {}", s.area());
+}
+```
+
+For `HasArea` and `Square`, we just declare a type parameter `T` and replace
+`f64` with it. The `impl` needs more involved modifications:
+
+```rust,ignore
```
Comment (steveklabnik) [inline src/doc/trpl/operators-and-overloading.md]: this should be just `ignore`, rust is implied

File: src/doc/trpl/traits.md [hunk 5]
```diff
@@ -47,8 +47,11 @@ As you can see, the `trait` block looks very similar to the `impl` block,
 but we don’t define a body, just a type signature. When we `impl` a trait,
 we use `impl Trait for Item`, rather than just `impl Item`.
 
-We can use traits to constrain our generics. Consider this function, which
-does not compile:
+## Traits for generic functions
```
Comment (steveklabnik) [inline src/doc/trpl/traits.md]: :confetti_ball: in general, this book can use more and more useful headings

Comment (steveklabnik) [inline src/doc/trpl/traits.md]: What about 'trait bounds for generic functions'?

File: src/doc/trpl/traits.md [hunk 6]
```diff
@@ -75,7 +78,7 @@ fn print_area<T: HasArea>(shape: T) {
 }
 ```
 
-The syntax `<T: HasArea>` means `any type that implements the HasArea trait`.
+The syntax `<T: HasArea>` means “any type that implements the `HasArea` trait”.
```
Comment (steveklabnik) [inline src/doc/trpl/traits.md]: period goes inside double quotes.

File: src/doc/trpl/traits.md [hunk 7]
```diff
@@ -152,6 +155,63 @@ We get a compile-time error:
 error: the trait `HasArea` is not implemented for the type `_` [E0277]
 ```
 
+## Traits for generic structs
+
+Your generic structs can also benefit from trait constraints. All you need to
+do is append the constraint when you declare type parameters. Here is a new
+type `Rectangle` and its operation `is_square`:
```
Comment (steveklabnik) [inline src/doc/trpl/traits.md]: can you make this `is_square()` please?

Comment (steveklabnik) [inline src/doc/trpl/traits.md]: Also it should be `Rectangle<T>`.

File: src/doc/trpl/traits.md [hunk 8]
```diff
@@ -152,6 +155,63 @@ We get a compile-time error:
...
+    };
+
+    assert!(r.is_square());
+
+    r.height = 42;
+    assert!(!r.is_square());
+}
+```
+
+`is_square` needs to check that the sides are equal, so the sides must be of a
```
Comment (steveklabnik) [inline src/doc/trpl/traits.md]: same here with the `()`s

File: src/doc/trpl/traits.md [hunk 9]
```diff
@@ -152,6 +155,63 @@ We get a compile-time error:
...
+type that implements the [`core::cmp::PartialEq`][PartialEq] trait:
+
+```rust,ignore
```
Comment (steveklabnik) [inline src/doc/trpl/traits.md]: same here, just `ignore`

**Comment by steveklabnik** [comment]
Looking great! Just have some nits. Would you mind squashing this too please?

**Comment by lastorset** [comment]
@steveklabnik Thanks for the review! I'll fix the issues you found.

Do you want it all current commits squashed into one commit (or two: new material and edits), or just my fixes to your review comments squashed into the current commits? Do I need to submit a new PR for the squashed branch, or can I force push to this one? (Haven't done this on GitHub much, I'm used to Critic. :))

**Comment by lastorset** [comment]
Documentation for [another project](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) implies that I can force push into this PR. So that's one question answered. (I do wonder what happens to the review comments. Probably GitHub is smart enough to keep them and the original commit objects around.)

**Comment by lastorset** [comment]
It was unclear how exactly I should squash, but I think a reasonable interpetration would be to keep the new material in separate commits and squash down all the copyedits. I push a [separate branch](https://github.com/rust-lang/rust/compare/master...lastorset:trait-operator-impl-unsquashed?expand=1) where I amended the original commits; you can merge that if you want more cohesive commits.

**Comment by steveklabnik** [comment]
hey sorry i missed this! You can squash into one or into more logical commits, like you've done here. thanks so much!

**Comment by lastorset** [comment]
No problem, thank you too!

**Pull request merged**